File: singleton/conectDB.js
// El Singleton (Implementación con Clase y Método Estático)
class GestorConexion {
    // 1. Almacén de la instancia
    static #instancia = null;

    // 2. Constructor privado (no se puede llamar directamente)
    constructor() {
        if (GestorConexion.#instancia) {
            // Lanza un error o retorna la instancia existente para forzar el Singleton
            throw new Error("Solo se puede crear una instancia de GestorConexion.");
        }
        this.estadoConexion = "Desconectado";
        console.log("Gestor de Conexión Creado (¡Solo una vez!)");
    }

    // 3. Método estático para obtener la instancia
    static getInstancia() {
        if (!GestorConexion.#instancia) {
            // Crea la instancia solo si no existe
            GestorConexion.#instancia = new GestorConexion();
        }
        return GestorConexion.#instancia;
    }

    conectar() {
        this.estadoConexion = "Conectado";
        console.log(`Conexión establecida. Estado: ${this.estadoConexion}`);
    }
}

// Uso:
/*
try {
    const c1 = GestorConexion.getInstancia();
    c1.conectar();

    const c2 = GestorConexion.getInstancia();
    console.log(`c1 es igual a c2: ${c1 === c2}`); // True
    c2.conectar(); 

    // Intentar crear directamente fallará (si se usa la validación en el constructor)
    // const c3 = new GestorConexion(); 
} catch (e) {
    console.error(e.message);
}
*/

File: 01 - Singleton/conectDB.cs
// El Singleton
public sealed class GestorConexion
{
    // 1. Instancia estática privada (Lazy loading)
    private static readonly Lazy<GestorConexion> _lazyInstance =
        new Lazy<GestorConexion>(() => new GestorConexion());

    // 2. Propiedad estática para acceder a la única instancia
    public static GestorConexion GetInstancia => _lazyInstance.Value;

    // 3. Constructor privado para prevenir la instanciación externa
    private GestorConexion()
    {
        Console.WriteLine("Gestor de Conexión Creado (¡Solo una vez!)");
        this.EstadoConexion = "Desconectado";
    }

    public string EstadoConexion { get; private set; }

    public void Conectar()
    {
        this.EstadoConexion = "Conectado";
        Console.WriteLine($"Conexión establecida. Hash: {this.GetHashCode()}");
    }

    public void Desconectar()
    {
        this.EstadoConexion = "Desconectado";
        Console.WriteLine("Conexión finalizada.");
    }
}

// Uso:
/*
// Ambas variables apuntan a la misma única instancia
GestorConexion c1 = GestorConexion.GetInstancia;
c1.Conectar();

GestorConexion c2 = GestorConexion.GetInstancia;
Console.WriteLine($"c1 es igual a c2: {c1 == c2}"); // True
c2.Desconectar(); 
*/
